client/extensions/woocommerce/app/order/order-details/fee-dialog.js

```diff
@@ -0,0 +1,141 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { uniqueId } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import Dialog from 'components/dialog';
+import { editOrder } from 'woocommerce/state/ui/orders/actions';
+import FormLabel from 'components/forms/form-label';
+import FormTextInput from 'components/forms/form-text-input';
+import { getCurrencyFormatDecimal } from 'woocommerce/lib/currency';
+import { getOrderWithEdits } from 'woocommerce/state/ui/orders/selectors';
+import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
+import PriceInput from 'woocommerce/components/price-input';
+
+class OrderFeeDialog extends Component {
+	static propTypes = {
+		isVisible: PropTypes.bool.isRequired,
+		editOrder: PropTypes.func.isRequired,
+		order: PropTypes.shape( {
+			currency: PropTypes.string.isRequired,
+			id: PropTypes.number.isRequired,
+		} ),
+		siteId: PropTypes.number.isRequired,
+		translate: PropTypes.func.isRequired,
+	};
+
+	state = {
+		name: '',
+		total: 0,
+	};
+
+	componentWillUpdate( nextProps ) {
+		// Dialog is being closed, clear the state
+		if ( this.props.isVisible && ! nextProps.isVisible ) {
+			this.setState( {
+				name: '',
+				total: 0,
+			} );
+		}
+	}
+
+	handleChange = event => {
+		const value = event.target.value;
+		switch ( event.target.name ) {
+			case 'new_fee_name':
+				this.setState( { name: value } );
+				break;
+			case 'new_fee_total':
+				this.setState( { total: value > 0 ? value : 0 } );
+				break;
+		}
+	};
+
+	formatCurrencyInput = () => {
+		const { currency } = this.props.order;
+		this.setState( prevState => ( {
+			total: getCurrencyFormatDecimal( prevState.total, currency ),
+		} ) );
+	};
+
+	handleFeeSave = () => {
+		const { siteId, order } = this.props;
+		if ( siteId ) {
+			const { name, total } = this.state;
+			const feeLines = order.fee_lines || [];
+			const tempId = uniqueId( 'fee_' );
+			this.props.editOrder( siteId, {
+				id: order.id,
+				fee_lines: [ ...feeLines, { id: tempId, name, total } ],
+			} );
+		}
+		this.props.closeDialog();
+	};
+
+	render() {
+		const { closeDialog, isVisible, order, translate } = this.props;
+		const dialogClass = 'woocommerce order-details__dialog'; // eslint/css specificity hack
+
+		const canSave = this.state.name && this.state.total;
```

Comment:
Nit: You can enter a name (e.g. 'Blerg') and then a fractional cent fee (e.g. 0.001) and the Add Fee button will light up (but doesn't actually add the fee.) Maybe tie the button state to both name and a minimum fee of 0.01?